### PR TITLE
Use Spotless to auto check and reformat Java/Scala code

### PR DESCRIPTION
**Pull request by pan3793**
### What changes were proposed in this pull request?

This PR proposes to use spotless to check and auto format the Java/Scala code.

The Java code uses Google Style, and the Scala Code style is defined in `.scalafmt.conf`.

This PR also provides a GitHub workflow for checking, and a shell script `dev/reformat` for quickly fixing. 

### Why are the changes needed?

It replaces the checkstyle and scalastyle maven plugins which are only for checking but requires the developer to fix them one by one manually.

### How was this patch tested?

/cc @related-reviewer

/assign @main-reviewer

### CODE REVIEW

**Comment by pan3793** [comment]
cc @waitinfuture @FMX @AngersZhuuuu, please let me know what you think of this change.

File: client-spark/shuffle-manager-2/src/main/java/org/apache/spark/shuffle/rss/HashBasedShuffleWriter.java [hunk 1]
```diff
@@ -17,17 +17,14 @@
 
 package org.apache.spark.shuffle.rss;
 
+import com.aliyun.emr.rss.client.ShuffleClient;
+import com.aliyun.emr.rss.client.write.DataPusher;
+import com.aliyun.emr.rss.common.RssConf;
```
Comment (pan3793) [inline client-spark/shuffle-manager-2/src/main/java/org/apache/spark/shuffle/rss/HashBasedShuffleWriter.java:22]: tried to keep import orders, but failed, guess it does not compatible w/ google style

File: .scalafmt.conf [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+maxColumn = 100
+newlines.source = keep
+newlines.topLevelStatementBlankLines = []
+optIn.configStyleArguments = false
+rewrite.imports.groups = [
+  ["javax?\\..*"],
+  ["scala\\..*"],
+  ["(?!com\\.aliyun\\.emr\\.rss\\.).*"],
+  ["com\\.aliyun\\.emr\\.rss\\..*"]
+]
```
Comment (AngersZhuuuu) [inline .scalafmt.conf:26]: Seems can't keep `com.aliyun.emr.rss` in the end?

Comment (pan3793) [reply .scalafmt.conf:26]: it's only for scala files

File: client-spark/shuffle-manager-2/src/main/java/org/apache/spark/shuffle/rss/HashBasedShuffleWriter.java (hunk 1 above)
Comment (pan3793) [reply client-spark/shuffle-manager-2/src/main/java/org/apache/spark/shuffle/rss/HashBasedShuffleWriter.java:22]: ref: https://github.com/diffplug/spotless/tree/main/plugin-maven#java

File: .scalafmt.conf (hunk 2 above)
Comment (AngersZhuuuu) [reply .scalafmt.conf:26]: move to `dev/.scalafmt.conf`?

Comment (pan3793) [reply .scalafmt.conf:26]: prefer the root folder, to make IDEA pick it up automatically

File: client-spark/shuffle-manager-2/src/main/java/org/apache/spark/shuffle/rss/HashBasedShuffleWriter.java (hunk 1 above)
Comment (pan3793) [reply client-spark/shuffle-manager-2/src/main/java/org/apache/spark/shuffle/rss/HashBasedShuffleWriter.java:22]: I figured it out, I should define the `<importOrder>` part to the last step

**Review by waitinfuture — APPROVED** [review]
LGTM, thanks!

**Pull request merged**